### PR TITLE
tracing v2 implementation

### PR DESCRIPTION
**Pull request by PietroPasotti**
Tandem PR: https://github.com/canonical/charm-relation-interfaces/pull/136

### CODE REVIEW

**Comment by mmkay** [comment]
First of all I wanted to check how easy is it to switch to v2. I took prometheus as one of the charms that I recently updated tracing in. My branch is here: https://github.com/canonical/prometheus-k8s-operator/tree/tracing-v2 . Commands to deploy tempo and prometheus:

```
juju deploy --trust ./tempo-k8s_ubuntu-22.04-amd64.charm tempo --resource tempo-image=grafana/tempo:1.5.0
juju deploy --trust ./prometheus-k8s_ubuntu-20.04-amd64.charm --resource prometheus-image=ubuntu/prometheus:2-22.04 && juju integrate tempo prometheus-k8s:tracing
```

There might be a bit of a race condition, as although I initialized tracing using

```
self.tracing = TracingEndpointRequirer(self, protocols=["otlp_http"])
```

on the endpoint method:
```
@property  
def tempo(self) -> Optional[str]:  
    """Tempo endpoint for charm tracing."""  
    return self.tracing.get_endpoint("otlp_http")
```
I got this exception:
```
unit-prometheus-0: 17:06:42 WARNING unit.prometheus/0.juju-log <class '__main__.PrometheusCharm'>.<property object at 0x7f943cccd590> returned None; continuing with tracing DISABLED.
unit-prometheus-k8s-0: 17:07:09 INFO juju.worker.uniter awaiting error resolution for "install" hook
unit-prometheus-k8s-0: 17:07:10 INFO unit.prometheus-k8s/0.juju-log Running legacy hooks/install.
unit-prometheus-k8s-0: 17:07:10 ERROR unit.prometheus-k8s/0.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "./src/charm.py", line 1074, in <module>
    main(PrometheusCharm)
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/ops/main.py", line 444, in main
    charm = charm_class(framework)
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/lib/charms/tempo_k8s/v1/charm_tracing.py", line 319, in wrap_init
    tracing_endpoint = _get_tracing_endpoint(tracing_endpoint_getter, self, charm)
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/lib/charms/tempo_k8s/v1/charm_tracing.py", line 241, in _get_tracing_endpoint
    tracing_endpoint = tracing_endpoint_getter.__get__(self)
  File "./src/charm.py", line 1065, in tempo
    return self.tracing.get_endpoint("otlp_http")
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/lib/charms/tempo_k8s/v2/tracing.py", line 744, in get_endpoint
    raise ProtocolNotRequestedError(protocol, relation)
charms.tempo_k8s.v2.tracing.ProtocolNotRequestedError: ('otlp_http', None)
```

After I added `if self.tracing.is_ready():` I was still getting connection issues.

```
unit-prometheus-k8s-0: 17:36:03 ERROR unit.prometheus-k8s/0.juju-log Exception while exporting Span batch.
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/urllib3/connection.py", line 198, in _new_conn
    sock = connection.create_connection(
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/urllib3/util/connection.py", line 85, in create_connection
    raise err
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/urllib3/util/connection.py", line 73, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/urllib3/connectionpool.py", line 793, in urlopen
    response = self._make_request(
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/urllib3/connectionpool.py", line 496, in _make_request
    conn.request(
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/urllib3/connection.py", line 400, in request
    self.endheaders()
  File "/usr/lib/python3.8/http/client.py", line 1251, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1011, in _send_output
    self.send(msg)
  File "/usr/lib/python3.8/http/client.py", line 951, in send
    self.connect()
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/urllib3/connection.py", line 238, in connect
    self.sock = self._new_conn()
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/urllib3/connection.py", line 213, in _new_conn
    raise NewConnectionError(
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7fd2342b4640>: Failed to establish a new connection: [Errno 111] Connection refused

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/requests/adapters.py", line 486, in send
    resp = conn.urlopen(
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/urllib3/connectionpool.py", line 847, in urlopen
    retries = retries.increment(
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/urllib3/util/retry.py", line 515, in increment
    raise MaxRetryError(_pool, url, reason) from reason  # type: ignore[arg-type]
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='tempo-0.tempo-endpoints.cos.svc.cluster.local', port=4318): Max retries exceeded with url: /v1/traces (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fd2342b4640>: Failed to establish a new connection: [Errno 111] Connection refused'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/opentelemetry/sdk/trace/export/__init__.py", line 368, in _export_batch
    self.span_exporter.export(self.spans_list[:idx])  # type: ignore
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/opentelemetry/exporter/otlp/proto/http/trace_exporter/__init__.py", line 145, in export
    resp = self._export(serialized_data)
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/opentelemetry/exporter/otlp/proto/http/trace_exporter/__init__.py", line 114, in _export
    return self._session.post(
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/requests/sessions.py", line 637, in post
    return self.request("POST", url, data=data, json=json, **kwargs)
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/requests/adapters.py", line 519, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='tempo-0.tempo-endpoints.cos.svc.cluster.local', port=4318): Max retries exceeded with url: /v1/traces (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fd2342b4640>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

Relation data from `jhack show-relation prometheus-k8s tempo:tracing`:
```
│                                                                                                  │
│ ╭───────────────────────────────────────── locals ─────────────────────────────────────────╮     │
│ │  local_endpoint = 'tracing'                                                              │     │
│ │         matches = []                                                                     │     │
│ │             obj = 'tempo:tracing'                                                        │     │
│ │       other_obj = 'prometheus-k8s'                                                       │     │
│ │        relation = Relation(                                                              │     │
│ │                   │   provider='tempo',                                                  │     │
│ │                   │   provider_endpoint='tracing',                                       │     │
│ │                   │   requirer='prometheus-k8s',                                         │     │
│ │                   │   requirer_endpoint='tracing',                                       │     │
│ │                   │   interface='tracing',                                               │     │
│ │                   │   raw_type='regular'                                                 │     │
│ │                   )                                                                      │     │
│ │       relations = [                                                                      │     │
│ │                   │   {                                                                  │     │
│ │                   │   │   'relation-id': 1,                                              │     │
│ │                   │   │   'endpoint': 'tracing',                                         │     │
│ │                   │   │   'related-endpoint': 'tracing',                                 │     │
│ │                   │   │   'application-data': {                                          │     │
│ │                   │   │   │   'host': '"tempo-0.tempo-endpoints.cos.svc.cluster.local"', │     │
│ │                   │   │   │   'receivers': '[{"protocol": "otlp_http", "port": 4318}]'   │     │
│ │                   │   │   },                                                             │     │
│ │                   │   │   'local-unit': {'in-scope': False, 'data': None},               │     │
│ │                   │   │   'related-units': {                                             │     │
│ │                   │   │   │   'tempo/0': {                                               │     │
│ │                   │   │   │   │   'in-scope': True,                                      │     │
│ │                   │   │   │   │   'data': {                                              │     │
│ │                   │   │   │   │   │   'egress-subnets': '10.152.183.113/32',             │     │
│ │                   │   │   │   │   │   'ingress-address': '10.152.183.113',               │     │
│ │                   │   │   │   │   │   'private-address': '10.152.183.113'                │     │
│ │                   │   │   │   │   }                                                      │     │
│ │                   │   │   │   }                                                          │     │
│ │                   │   │   }                                                              │     │
│ │                   │   }                                                                  │     │
│ │                   ]                                                                      │     │
│ │ remote_endpoint = None                                                                   │     │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────╯     │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯

```
Maybe there's an issue with port opening? It doesn't look like `4318` is exposed by the service:
```
$ kubectl get services -A
NAMESPACE             NAME                       TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                  AGE
default               kubernetes                 ClusterIP   10.152.183.1     <none>        443/TCP                  31h
kube-system           kube-dns                   ClusterIP   10.152.183.10    <none>        53/UDP,53/TCP,9153/TCP   31h
metallb-system        webhook-service            ClusterIP   10.152.183.24    <none>        443/TCP                  31h
controller-microk8s   controller-service         ClusterIP   10.152.183.46    <none>        17070/TCP                30h
controller-microk8s   modeloperator              ClusterIP   10.152.183.209   <none>        17071/TCP                30h
pietro                modeloperator              ClusterIP   10.152.183.162   <none>        17071/TCP                6m34s
pietro                tempo                      ClusterIP   10.152.183.63    <none>        65535/TCP                6m12s
pietro                tempo-endpoints            ClusterIP   None             <none>        <none>                   6m11s
pietro                prometheus-k8s-endpoints   ClusterIP   None             <none>        <none>                   3m42s
pietro                prometheus-k8s             ClusterIP   10.152.183.115   <none>        9090/TCP                 3m43s
```
while the one from edge seems to expose a ton of ports:
```
pietro                tempo-k8s                  ClusterIP   10.152.183.99    <none>        3200/TCP,4317/TCP,4318/TCP,9411/TCP,14268/TCP,14250/TCP   40s
```

File: lib/charms/tempo_k8s/v2/tracing.py [hunk 1]
```diff
@@ -0,0 +1,747 @@
...
+        self.expected_relation_interface = expected_relation_role
+        self.actual_relation_role = actual_relation_role
+        self.message = "The '{}' relation has role '{}' rather than the expected '{}'".format(
+            relation_name, repr(actual_relation_role), repr(expected_relation_role)
+        )
+
+        super().__init__(self.message)
+
+
+def _validate_relation_by_interface_and_direction(
```
Comment (mmkay) [inline lib/charms/tempo_k8s/v2/tracing.py:334]: I'm not sure if I get the relation of this relation validation wtih tracing. I see it's being used to verify if relations/integrations are correct. However it seems fairly generic. How complex in the current charm lib ecosystem would it be to extract it into something common that could be shared in various libraries? I see it's already copied around quite a lot:

https://github.com/search?q=org%3Acanonical%20_validate_relation_by_interface_and_direction(&type=code

File: lib/charms/tempo_k8s/v2/tracing.py [hunk 2]
```diff
@@ -0,0 +1,747 @@
...
+
+        self._charm = charm
+        self._relation_name = relation_name
+
+        events = self._charm.on[self._relation_name]
+        self.framework.observe(events.relation_changed, self._on_tracing_relation_changed)
+        self.framework.observe(events.relation_broken, self._on_tracing_relation_broken)
+
+        if protocols is not None:
+            self.request_protocols(protocols)
```
Comment (mmkay) [inline lib/charms/tempo_k8s/v2/tracing.py:612]: if `protocols` is `None` -> current default value, we don't enter this code branch. However, if it's an empty list, `request_protocols` activates all supported receivers. We should stay consistent.

File: lib/charms/tempo_k8s/v2/tracing.py [hunk 3]
```diff
@@ -0,0 +1,747 @@
...
+            # "ingesters" Might be populated if the provider sees a v1 relation before a v2 requirer has had time to
+            # publish the 'receivers' list. This will make Tempo incorrectly assume that this is a v1
+            # relation, and act accordingly. Later, when the requirer publishes the requested receivers,
+            # tempo will be able to course-correct.
+            if "ingesters" in databag:
+                del databag["ingesters"]
+            TracingProviderAppData.load(databag)
+
+        except (json.JSONDecodeError, pydantic.ValidationError, DataValidationError):
+            logger.info(f"failed validating relation data for {relation}")
```
Comment (mmkay) [inline lib/charms/tempo_k8s/v2/tracing.py:686]: We catch a validation error, but don't tell the user what was wrong with relation data. Could we log the exception traceback when charm is set to debug?

File: lib/charms/traefik_k8s/v2/ingress.py [hunk 4]
```diff
@@ -72,69 +72,146 @@ def _on_ingress_revoked(self, event: IngressPerAppRevokedEvent):
...
-PYDEPS = ["pydantic<2.0"]
+PYDEPS = ["pydantic"]
 
 DEFAULT_RELATION_NAME = "ingress"
 RELATION_INTERFACE = "ingress"
 
 log = logging.getLogger(__name__)
 BUILTIN_JUJU_KEYS = {"ingress-address", "private-address", "egress-subnets"}
 
+if pydantic.version.VERSION.split(".") <= ["2"]:
```
Comment (mmkay) [inline lib/charms/traefik_k8s/v2/ingress.py:85]: :heart:

File: src/tempo.py [hunk 5]
```diff
@@ -20,38 +20,48 @@ class Tempo:
...
 
-    def __init__(
-        self, port: int = 3200, grpc_listen_port: int = 9096, local_host: str = "0.0.0.0"
-    ):
-        self.tempo_port = port
-
+    # todo make configurable?
+    receiver_ports: Dict[ReceiverProtocol, int] = {
+        "zipkin": 9411,
+        "kafka": 42,  # TODO:
```
Comment (mmkay) [inline src/tempo.py:26]: how about we make kafka/opencensus/thrift compact/binary a separate issue to avoid too much scope creep?

File: src/tempo.py [hunk 6]
```diff
@@ -163,3 +158,36 @@ def is_ready(self):
...
+        config = {}
+
+        if "zipkin" in receivers_set:
+            config["zipkin"] = None
+        if "opencensus" in receivers_set:
+            config["opencensus"] = None
+
+        otlp_config = {}
+        if "otlp_http" in receivers_set:
+            otlp_config["http"] = None
```
Comment (mmkay) [inline src/tempo.py:175]: This might still need to be filled? Unless I miss something which is possible at this point ;)

File: tests/integration/tester/lib/charms/tempo_k8s/v1/charm_tracing.py
Comment (mmkay) [inline tests/integration/tester/lib/charms/tempo_k8s/v1/charm_tracing.py:1]: I think I remember a discussion about not adding these libraries into the repo so that they get recreated at every test run? I'm okay with keeping them here if it doesn't affect anything but this will mean more duplication in the repo.

**Review by mmkay — CHANGES_REQUESTED** [review]
Nice work! There's still some teething issues and local manual tests weren't fully successful yet but I think it's moving in a great direction.

File: lib/charms/tempo_k8s/v2/tracing.py [hunk 7]
```diff
@@ -0,0 +1,747 @@
+# Copyright 2022 Pietro Pasotti
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v2/tracing.py:1]: :'D

[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v2/tracing.py:1]: I suspect this is an org-wide issue

File: lib/charms/tempo_k8s/v2/tracing.py (hunk 3 above)
Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v2/tracing.py:686]: 90% of the cases this is due to the databag being empty, we could special-case that and if it's nonempty make sure we surface something.

File: lib/charms/tempo_k8s/v1/charm_tracing.py [hunk 8]
```diff
@@ -534,6 +535,8 @@ def wrapped_function(*args, **kwargs):  # type: ignore
         name = getattr(callable, "__qualname__", getattr(callable, "__name__", str(callable)))
         with _span(f"{'(static) ' if static else ''}{qualifier} call: {name}"):  # type: ignore
             if static:
+                # fixme: do we or don't we need [1:]?
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v1/charm_tracing.py:538]: yeah I'm not super clear about it either. The decorator doesn't always play nice with `@staticmethod`s. Sometimes it will receive 'self', sometimes it won't. Depending on the test run I have to add that or remove it, I'm not yet sure why.

File: lib/charms/tempo_k8s/v2/tracing.py [hunk 9]
```diff
@@ -0,0 +1,747 @@
...
+                    logger.warning("unused ssl argument - was the right protocol called?")
+                return f"{ep.host}:{receiver.port}"
+            if ssl:
+                return f"https://{ep.host}:{receiver.port}"
+            return f"http://{ep.host}:{receiver.port}"
+        except StopIteration:
+            logger.error(f"no receiver found with protocol={protocol!r}")
+            return None
+
+    def get_endpoint(
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v2/tracing.py:732]: very good point. Will add them back.

File: lib/charms/tempo_k8s/v2/tracing.py [hunk 10]
```diff
@@ -0,0 +1,747 @@
...
+
+
+class AmbiguousRelationUsageError(TracingError):
+    """Raised when one wrongly assumes that there can only be one relation on an endpoint."""
+
+
+class DatabagModel(BaseModel):
+    """Base databag model."""
+
+    model_config = ConfigDict(
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v2/tracing.py:160]: yeah I wanted to have the blueprint on how to do that ready in traefik, then copy/paste the databagmodel here
Will do that now.

[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v2/tracing.py:160]: traefik is kind of the 'single source of truth for DataBagModel'

[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v2/tracing.py:160]: unofficially

File: lib/charms/tempo_k8s/v2/tracing.py (hunk 1 above)
Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v2/tracing.py:334]: true, but not an issue for this PR. sounds like a feature request for `ops`.

File: lib/charms/tempo_k8s/v2/tracing.py (hunk 2 above)
Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v2/tracing.py:612]: good catch.

File: src/tempo.py (hunk 6 above)
Comment (PietroPasotti) [reply src/tempo.py:175]: mmm, what do you mean? as far as I can tell from https://github.com/open-telemetry/opentelemetry-collector/tree/overlord/receiver , the minimal config is `receiver-name: null`.

File: tests/integration/tester/lib/charms/tempo_k8s/v1/charm_tracing.py
Comment (PietroPasotti) [reply tests/integration/tester/lib/charms/tempo_k8s/v1/charm_tracing.py:1]: yup, my mistake

File: src/tempo.py (hunk 6 above)
Comment (mmkay) [reply src/tempo.py:175]: Might've been under the impression that this is the place where we enter port numbers into the config. If that's just the structure, my comment can be ignored.

File: lib/charms/tempo_k8s/v2/tracing.py [hunk 11]
```diff
@@ -0,0 +1,747 @@
...
+class TracingEndpointRequirer(Object):
+    """A tracing endpoint for Tempo."""
+
+    on = TracingEndpointRequirerEvents()  # type: ignore
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        protocols: Optional[List[ReceiverProtocol]] = None,
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v2/tracing.py:569]: to be fully BC (as in: replace v1 with v2 in the imports and all 'works') we should have 5 protocols in there or so.
IMHO it's fine if we are not in this case, as the whole point of using v2 is to be able to specify the protocols. If you don't need that, just keep using v1.

Something we could consider is making the protocols arg mandatory or removing it altogether.
I made it mandatory so that to avoid races one could call `request_protocols` at any time after init. 
But we could also simply force the user to manually call request_protocols, at the risk that they can more easily forget to do so.

[the remark this answers is not in the document]
Comment (mmkay) [reply lib/charms/tempo_k8s/v2/tracing.py:569]: I am okay with keeping it a mandatory argument - especially at this stage in charm_tracing adoption. IMO forcing to call `request_protocols` manually would be making the UX of adding charm_tracing worse.

File: src/tempo.py (hunk 6 above)
Comment (PietroPasotti) [reply src/tempo.py]: might be a difference between tempo v1 and v2. Couldn't find a clear answer in the docs.
For now it seems to be working with cos-lite so we'll fix and validate this in the next PR

**Review by mmkay — APPROVED** [review]
The implementation looks good - please remember about bumping `charm_tracing` :)

File: src/tempo.py [hunk 12]
```diff
@@ -167,8 +171,14 @@ def is_ready(self):
...
-    def _build_receivers_config(self, receivers: Sequence[ReceiverProtocol]):  # noqa: C901
-        receivers_set = set(receivers)  # convert to set for faster lookup
+    def _build_receivers_config(self, remote_receivers: Sequence[ReceiverProtocol]):  # noqa: C901
+        # remote_receivers: the receivers we have to enable because the requirers we're related to
+        # intend to use them
+        # self._enabled_receivers: receivers we have to enable because *this charm* will use them.
+        receivers_set = set(remote_receivers)
+        receivers_set.update(self._enabled_receivers)
+        if not receivers_set:
+            logger.warning("No receivers set. Tempo will be up but not functional.")
```
Comment (mmkay) [inline src/tempo.py:181]: IMO if there is/was something like charm metrics that could be a reason for an alert.

Comment (PietroPasotti) [reply src/tempo.py:181]: yeah on the operator side I'm not sure if it would be reasonable to put the charm in blocked
fact is, until you relate it, tempo has no reason to even start if not to monitor itself (via charm_tracing)

and in practice, thanks to charm_tracing, there will always be at least `otlp_http` in there so this should never happen.

File: lib/charms/grafana_k8s/v0/grafana_source.py [hunk 13]
```diff
@@ -724,9 +724,23 @@ def _set_default_data(self) -> None:
...
+            return
+
+        peers.data[self._charm.app][key] = json.dumps(data)  # type: ignore[attr-defined]
 
     def get_peer_data(self, key: str) -> Any:
         """Retrieve information from the peer data bucket instead of `StoredState`."""
+        peers = self._charm.peers  # type: ignore[attr-defined]
+        if not peers:
+            # https://bugs.launchpad.net/juju/+bug/1998282
+            logger.warning(
```
[the remark this answers is not in the document]
Comment (mmkay) [reply lib/charms/grafana_k8s/v0/grafana_source.py:740]: silly me then ;)
